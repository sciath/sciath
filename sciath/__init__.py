""" SciATH: Scientific Application Test Harness """
from sciath._sciath_io import NamedColors

__version__ = (0, 7, 0)

# A default set of colors
SCIATH_COLORS = NamedColors()
