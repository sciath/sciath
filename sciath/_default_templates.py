""" Generate default templates for Launcher batch systems """

import os

CONTENTS_LOCAL = r"""#!/usr/bin/env sh

$SCIATH_TASK_MPI_RUN \
$SCIATH_TASK_COMMAND 1>>$SCIATH_JOB_STDOUT 2>>$SCIATH_JOB_STDERR; printf "$?\n" >> $SCIATH_JOB_EXITCODE

touch $SCIATH_JOB_COMPLETE
"""

CONTENTS_LSF = r"""#!/bin/sh

#BSUB -J $SCIATH_JOB_NAME
#BSUB -o $SCIATH_JOB_STDOUT
#BSUB -e $SCIATH_JOB_STDERR
#BSUB -n $SCIATH_JOB_MAX_RANKS_OR_REMOVE_LINE
#BSUB -W $SCIATH_JOB_WALLTIME_HM_OR_REMOVE_LINE
#BSUB -q $SCIATH_QUEUE_OR_REMOVE_LINE

$SCIATH_TASK_MPI_RUN \
$SCIATH_TASK_COMMAND; printf "$?\n" >> $SCIATH_JOB_EXITCODE

touch $SCIATH_JOB_COMPLETE
"""

CONTENTS_SLURM = r"""#!/bin/bash -l

#SBATCH --job-name=$SCIATH_JOB_NAME
#SBATCH --output=$SCIATH_JOB_STDOUT
#SBATCH --error=$SCIATH_JOB_STDERR
#SBATCH --ntasks=$SCIATH_JOB_MAX_RANKS_OR_REMOVE_LINE
#SBATCH --time=$SCIATH_JOB_WALLTIME_HMS_OR_REMOVE_LINE
#SBATCH --account=$SCIATH_ACCOUNT_OR_REMOVE_LINE
#SBATCH --partition=$SCIATH_QUEUE_OR_REMOVE_LINE

$SCIATH_TASK_MPI_RUN \
$SCIATH_TASK_COMMAND; printf "$?\n" >> $SCIATH_JOB_EXITCODE

touch $SCIATH_JOB_COMPLETE
"""


def _generate_default_template(system_type, output_path=".", filename=None):
    if system_type == "local":
        contents = CONTENTS_LOCAL
        if filename is None:
            filename = "SciATH_template_local.sh"
    elif system_type == "lsf":
        contents = CONTENTS_LSF
        if filename is None:
            filename = "SciATH_template_lsf.lsf"
    elif system_type == "slurm":
        contents = CONTENTS_SLURM
        if filename is None:
            filename = "SciATH_template_slurm.sbatch"
    else:
        raise Exception("Unrecognized batch system type %s" % system_type)

    with open(os.path.join(output_path, filename), "w") as file:
        file.write(contents)
    return filename
