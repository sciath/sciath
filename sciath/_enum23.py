# Defines a class which behaves like an Enum (Python 3.4+)
# But also supports some operations with earlier versions of Python
# It is used just like Enum, except that values must be assigned
# Using Enum23Value(), not directly, for example
#
# class MyEnum(Enum23):
#    FOO = Enum23Value('foo')
#
# If Python <3.4 support is dropped, Enum23 can be directly replaced
# with Enum, and Enum23Value() can be removed

import sys

if sys.version_info[0] >= 3 and sys.version_info[1] >= 4:
    from enum import Enum

    def Enum23Value(x):
        return x

    Enum23 = Enum
else:
    class Enum23Value:
        """ A wrapper class which allows one to access a value field """
        def __init__(self, value): self.__dict__.update(value=value)

    Enum23 = object
