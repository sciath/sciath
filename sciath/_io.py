from __future__ import print_function

import os
import sys

def py23input(prompt) :
    if sys.version_info[0] == 2 :
        v = raw_input(prompt)
    else :
        v = input(prompt)
    return(v)


def _remove_file_if_it_exists(filename):
    try:
        os.remove(filename)
    except OSError:
        pass


def command_join(command):
    """ Convert a command (as would go to subprocess.run()) to a copy-pasteable string

    Do something similar to shlex.join (Python 3.8+), attempting to quote arguments
    that contain spaces, and escape newlines in the result. """

    joined = ' '.join(["'"+term+"'" if ' ' in term else term for term in command])
    joined = joined.replace('\n','\\n')
    return joined

class NamedColors:
    def __init__(self):
        self.set_colors()

    def set_colors(self,use_bash=True):
        self.HEADER    = '\033[35m' if use_bash else ''
        self.SUBHEADER = '\033[36m' if use_bash else ''
        self.OK        = '\033[32m' if use_bash else ''
        self.WARNING   = '\033[93m' if use_bash else ''
        self.FAIL      = '\033[91m' if use_bash else ''
        self.ENDC      = '\033[0m'  if use_bash else ''

# two space tab for formatted print statements
tab = '  '
