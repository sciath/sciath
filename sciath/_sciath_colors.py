""" SciATH colors for printing """


class NamedColors:  #pylint: disable=too-few-public-methods
    """ Color codes to use for SciATH output """

    def __init__(self):
        self.set_colors()

    def set_colors(self, use_bash=True):
        """ Set color codes, based on whether or not bash colors are used """
        self.header = '\033[35m' if use_bash else ''
        self.subheader = '\033[36m' if use_bash else ''
        self.okay = '\033[32m' if use_bash else ''
        self.warning = '\033[33m' if use_bash else ''
        self.error = '\033[91m' if use_bash else ''
        self.fail = '\033[91m' if use_bash else ''
        self.endc = '\033[0m' if use_bash else ''
