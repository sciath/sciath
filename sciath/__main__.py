import sciath.harness

def main():
    sciath.harness.Harness().run_from_args()

if __name__ == '__main__':
    main()
