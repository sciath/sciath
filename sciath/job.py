""" SciATH Job class """
from __future__ import print_function

import sciath.task


class Job:
    r""":class:`Job` describes a ordered list of :class:`Task`\s.

    It describes the :class:`Task`\s themselves, not information about any
    particular "run" (handled within :class:`Harness`) or how one might
    interpret the results of such a run (handled by :class:`Test`).  A
    :class:`Launcher` object executes the :class:`Task`\s described by a :class:`Job`.

    Data include

    * A name
    * An ordered list of :class:`Task`\s

    It also defines how the name may be converted to various standard filenames.

    """

    _default_job_name = 'job'

    def __init__(self, task_or_tasks, name=None):
        if name is None:
            self.name = Job._default_job_name
            self.named_by_default = True
        else:
            self.name = name.replace(' ', '_')
            self.named_by_default = False

        if isinstance(task_or_tasks, sciath.task.Task):
            self.tasks = [task_or_tasks]
        elif isinstance(task_or_tasks, list):
            for task in task_or_tasks:
                if not isinstance(task, sciath.task.Task):
                    raise Exception('A Task or list of Tasks was expected')
                self.tasks = task_or_tasks
        else:
            raise Exception('A Task or list of Tasks was expected')

    def create_execute_command(self):
        """
        Returns a list containing (command, resource) tuples for the Job.
        """
        return [task.create_execute_command() for task in self.tasks]

    @property
    def complete_filename(self):
        """ Returns a filename for a sentinel signifying completion """
        return '.%s.complete' % self.name

    @property
    def exitcode_filename(self):
        """ Returns a filename to use for exit codes """
        return self.name + '.exitcode'

    @property
    def launched_filename(self):
        """ Returns a filename for a sentinel signifying launching """
        return '.%s.launched' % self.name

    @property
    def stdout_filename(self):
        """ Returns a filename to use for stdout """
        return self.name + '.stdout'

    @property
    def stderr_filename(self):
        """ Returns a filename to use for stderr """
        return self.name + '.stderr'

    # This can be removed once the transition to using templates is complete
    def get_max_resources(self):
        """ Returns a dict() defining the maximum required counts / values """

        # Use the first Task to determine which resources to consider
        max_resources = {}
        for key, value in self.tasks[0].resources.items():
            max_resources[key] = value

        for task in self.tasks[1:]:
            for key, value in task.resources.items():
                if value > max_resources[key]:
                    max_resources[key] = value
        return max_resources

    def resource_max(self, key):
        """ Returns the maximum value of a resource over all Tasks

            Returns None if no Tasks define that resource.
        """
        maximum = None
        for task in self.tasks:
            if key in task.resources:
                current = task.resources[key]
                if maximum is None:
                    maximum = current
                else:
                    maximum = max(maximum, current)
        return maximum

    def number_tasks(self):
        """ Returns the number of tasks within the Job """
        return len(self.tasks)

    def total_wall_time(self):
        """ Returns the total wall time required for all Tasks

            Returns None if any task does not have a defined wall time.
        """
        total_wall_time = 0
        for task in self.tasks:
            if task.wall_time is None:
                return None
            total_wall_time += task.wall_time
        return total_wall_time
