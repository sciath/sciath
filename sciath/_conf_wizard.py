""" Interactive tool for populating Launcher configuration information """

import os
import shlex
import shutil

from sciath._sciath_io import py23input
from sciath._default_templates import _generate_default_template
from sciath.utility import DotDict

DEFAULTS_LOCAL = DotDict()
DEFAULTS_LOCAL.job_submission_command = ("sh",)
DEFAULTS_LOCAL.blocking = True
DEFAULTS_LOCAL.has_job_level_ranks = False

DEFAULTS_LSF = DotDict()
DEFAULTS_LSF.job_submission_command = (
    "sh",
    "-c",
    "bsub < $0",
)  # This allows "<".
DEFAULTS_LSF.blocking = False
DEFAULTS_LSF.has_job_level_ranks = True
DEFAULTS_LSF.mpi_launch = "mpirun"

DEFAULTS_SLURM = DotDict()
DEFAULTS_SLURM.job_submission_command = ("sbatch",)
DEFAULTS_SLURM.blocking = False
DEFAULTS_SLURM.has_job_level_ranks = True
DEFAULTS_SLURM.mpi_launch = "srun -n <ranks>"

DEFAULTS = {
    "local": DEFAULTS_LOCAL,
    "lsf": DEFAULTS_LSF,
    "slurm": DEFAULTS_SLURM,
}


def _query_user(description="",
                options=None,
                examples=None,
                default=None,
                validate=None):
    valid = False
    while not valid:
        options_string = "" if options is None else " <%s>" % ", ".join(options)

        print("%s%s" % (description, options_string))
        if examples is not None:
            print("Examples:")
            for example in examples:
                print(example)

        if default is not None:
            print("Hit enter to accept default [%s]" % default)

        result = py23input("> ")
        result = os.path.expandvars(result)
        result = os.path.expanduser(result)

        if result == "":
            result = default

        valid = options is None or result in options
        if validate is not None and valid:
            valid = validate(result)

    return result


def _launcher_interactive_configure(launcher):  #pylint: disable=too-many-statements, too-many-branches
    """ Create a new configuration file from user input """
    print("--------------------------------------------------------------")
    print("Creating new configuration file %s" % launcher.conf_filename)

    system_type_options = list(DEFAULTS.keys()) + ["generic"]
    system_type = _query_user("\n[?] Batch queuing system type",
                              options=system_type_options,
                              default="local")

    if system_type in DEFAULTS:
        launcher.job_submission_command = list(
            DEFAULTS[system_type].job_submission_command)
    else:
        launcher.job_submission_command = shlex.split(
            _query_user("\n[?] job submission command",
                        validate=lambda command: command is not None))

    if system_type in DEFAULTS:
        launcher.blocking = DEFAULTS[system_type].blocking
    else:
        launcher.blocking = _query_user("\n[?] blocking",
                                        options=["True", "False"]) == "True"

    if system_type in DEFAULTS:
        launcher.has_job_level_ranks = DEFAULTS[system_type].has_job_level_ranks
    else:
        launcher.has_job_level_ranks = _query_user(
            "\n[?] supports job-level ranks specification",
            options=["True", "False"]) == "True"

    mpi_launch_examples = [
        "  No MPI Required           : none",
        "  Local Machine (mpirun)    : mpirun -np <ranks>",
        "  Local Machine (mpiexec)   : mpiexec -np <ranks>",
        "  SLURM w/ aprun            : aprun -B",
        "  Native SLURM              : srun -n <ranks>",
        "  LSF (Euler)               : mpirun",
    ]
    petsc_dir = os.getenv("PETSC_DIR")
    petsc_arch = os.getenv("PETSC_ARCH")
    if petsc_dir and petsc_arch:
        mpi_launch_examples.append(
            "  Current PETSc MPI wrapper : %s -n <ranks>" %
            os.path.join(petsc_dir, petsc_arch, "bin", "mpiexec"))
    else:
        mpi_launch_examples.append(
            "  Example PETSc MPI wrapper : /users/myname/petsc/arch-xxx/bin/mpiexec -n <ranks>"
        )

    if system_type in DEFAULTS and "mpi_launch" in DEFAULTS[system_type]:
        mpi_launch_default = DEFAULTS[system_type].mpi_launch
    else:
        mpi_launch_default = "none"
    mpi_launch = _query_user(
        "\n[?] MPI launch command, including <ranks> for number of ranks",
        examples=mpi_launch_examples,
        default=mpi_launch_default,
        validate=launcher.is_valid_mpi_launch)

    launcher.set_mpi_launch(mpi_launch)

    launcher.account_name = _query_user("\n[?] Account to charge", default="")

    launcher.queue_name = _query_user("\n[?] Name of queue to submit tests to",
                                      default="")

    # Assume, dubiously, that defaults defined here have a default template defined,
    # that the user always wants to use.
    query_template = system_type not in DEFAULTS
    if query_template:
        template_source_filename = _query_user(
            "\n[?]Please select a file, to be copied here, to use as a Launcher template",
            validate=os.path.isfile,
        )
        template_filename = os.path.split(template_source_filename)[1]
        if not os.path.isfile(template_filename):
            shutil.copyfile(template_source_filename, template_filename)
        launcher.template_filename = template_filename
    else:
        launcher.template_filename = _generate_default_template(system_type)

    launcher.write_definition(launcher.conf_filename)

    print("")
    print("** The template for batch submission files is  %s" %
          launcher.template_filename)
    print("** You may modify it if desired, but it will be overwritten")
    print("** if you re-configure.")
    print("")
    print("** To change the config for your batch system, either")
    print("**  (i) delete the file %s  or" % launcher.conf_filename)
    print("** (ii) re-run with the command line arg --configure")
    print("--------------------------------------------------------------")
