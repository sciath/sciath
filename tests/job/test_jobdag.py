
from sciath.job import Job
from sciath.job import JobSequence
from sciath.job import JobDAG


def example0a():
  print('++++++++ example 0a ++++++++')
  
  jd1 = Job( 'echo \"dependent job 1\"', name = 'demo-job-1' )
  jd2 = Job( 'echo \"dependent job 2\"')

  
  # Example usage
  jA = JobDAG('echo \"job A\"',name='DAG with no deps.')
  jA.registerJob(jd1)
  jA.registerJob(jd2)
  jA.view()

def example1():
  print('++++++++ example 1 (DAG with no deps.) ++++++++')
  
  # Example usage
  jA = JobDAG('echo \"job A\"',name='jobA')

  user_dag = {'jobA': [None]}
  jA.insert(user_dag)

  er = jA.createExecuteCommand()
  print('Execute command + resources for jA')
  for i in er:
    print('  cmd =',i[0],'res =',i[1])

  print('============================================================================')
  jA.view()
  print('============================================================================')


def example2():
  print('++++++++ example 2 (DAG with one dep.) ++++++++')

  # Example usage
  jA = JobDAG('echo \"job A\"',name='jobA')
  jA.setResources(threads=27,ranks=40)

  jd1 = Job( 'echo \"dependent job 1\"', name = 'demo-depjob-1' )
  jd1.setResources(threads=101,ranks=101)

  jA.registerJob(jd1)


  user_dag = {'jobA': [ 'demo-depjob-1' ],
              'demo-depjob-1': [None]}

  jA.insert(user_dag)

  er = jA.createExecuteCommand()
  print('Execute command + resources for jA')
  for i in er:
    print('  cmd =',i[0],'res =',i[1])

  print('============================================================================')
  jA.view()
  print('============================================================================')


def example3():
  print('++++++++ example 3 (DAG with dependencies define by a non-trival graph) ++++++++')
  
  # Example usage
  jA = JobDAG('echo \"job A\"',name='jobA')
  jA.setResources(threads=27,ranks=40)

  jdb = Job( 'echo \"dependent job b\"', name = 'b' )
  jdc = Job( 'echo \"dependent job c\"', name = 'c' )
  jdd = Job( 'echo \"dependent job d\"', name = 'd' )
  jdd.setResources(threads=1,ranks=4)
  jde = Job( 'echo \"dependent job e\"', name = 'e' )
  jdf = Job( 'echo \"dependent job f\"', name = 'f' )
  jdg = Job( 'echo \"dependent job g\"', name = 'g' )
  jdh = Job( 'echo \"dependent job h\"', name = 'h' )

  jl = [ jdb, jdc, jdd, jde, jdf, jdg, jdh]
  for j in jl:
    jA.registerJob(j)


  dag = {'jobA': [ 'b', 'c', 'h'],
            'b': ('c','f'),
            'c': ('d', 'e', 'f'),
            'd': ('g'),
            'e': ('h'),
            'f': [None],
            'g': ([None]),
            'h': ([None])}

  jA.insert(dag)
  
  er = jA.createExecuteCommand()
  print('Execute command + resources for jA')
  for i in er:
    print('  cmd =',i[0],'res =',i[1])
  
  print('============================================================================')
  jA.view()
  print('============================================================================')



#example0a()

#example1()

#example2()

example3()







