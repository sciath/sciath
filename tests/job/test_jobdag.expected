++++++++ example 3 (DAG with dependencies define by a non-trival graph) ++++++++
('not found -> inserting name', 'b')
('not found -> inserting name', 'c')
('not found -> inserting name', 'd')
('not found -> inserting name', 'e')
('not found -> inserting name', 'f')
('not found -> inserting name', 'g')
('not found -> inserting name', 'h')
('[pass] The root vertex associated with parent job (name =', 'jobA', ') was found in the DAG.')
[pass] All DAG vertices were found in self.joblist.
('dag -> order =', ['g', 'e', 'd', 'f', 'h', 'c', 'b', 'jobA'])
Execute command + resources for jA
('  cmd =', 'echo "dependent job g"', 'res =', {'mpiranks': 1, 'threads': 1})
('  cmd =', 'echo "dependent job e"', 'res =', {'mpiranks': 1, 'threads': 1})
('  cmd =', 'echo "dependent job d"', 'res =', {'mpiranks': 4, 'threads': 1})
('  cmd =', 'echo "dependent job f"', 'res =', {'mpiranks': 1, 'threads': 1})
('  cmd =', 'echo "dependent job h"', 'res =', {'mpiranks': 1, 'threads': 1})
('  cmd =', 'echo "dependent job c"', 'res =', {'mpiranks': 1, 'threads': 1})
('  cmd =', 'echo "dependent job b"', 'res =', {'mpiranks': 1, 'threads': 1})
('  cmd =', 'echo "job A"', 'res =', {'mpiranks': 40, 'threads': 27})
============================================================================
('Job: Job name:', 'jobA')
('Command:', 'echo "job A"')
('Exit code success:', 0)
('Resources:', "{'mpiranks': 40, 'threads': 27}")
('Max. resources (incl. dependencies):', "{'mpiranks': 40, 'threads': 27}")
JobDAG:
[Registered jobs]
('  job', 0, 'key =', 'c')
('  job', 1, 'key =', 'b')
('  job', 2, 'key =', 'e')
('  job', 3, 'key =', 'd')
('  job', 4, 'key =', 'g')
('  job', 5, 'key =', 'f')
('  job', 6, 'key =', 'h')
[Provided DAG]
('  "g" ->', [None])
('  "c" ->', ('d', 'e', 'f'))
('  "b" ->', ('c', 'f'))
('  "e" ->', 'h')
('  "d" ->', 'g')
('  "jobA" ->', ['b', 'c', 'h'])
('  "f" ->', [None])
('  "h" ->', [None])
[Execution order]
('  order', 0, ':', 'g')
('  order', 1, ':', 'e')
('  order', 2, ':', 'd')
('  order', 3, ':', 'f')
('  order', 4, ':', 'h')
('  order', 5, ':', 'c')
('  order', 6, ':', 'b')
('  order', 7, ':', 'jobA')
============================================================================
